File: test/ipfs_accelerate_py/hardware/hardware_profile.py
"""
Hardware profile configuration for model acceleration.

This module provides a comprehensive hardware profile configuration system
that supports all hardware backends available in the IPFS acceleration system.
"""

from typing import Dict, Any, List, Optional, Union, Literal

class HardwareProfile:
    """
    Configuration profile for specific hardware backends.
    
    This class encapsulates all configuration options for specific hardware
    backends, providing a consistent interface for hardware-specific settings.
    """
    
    def __init__(
        self,
        backend: str = "auto",
        device_id: Union[int, str] = 0,
        memory_limit: Optional[Union[int, str]] = None,
        precision: str = "auto",
        optimization_level: Literal["default", "performance", "memory", "balanced"] = "default",
        quantization: Optional[Dict[str, Any]] = None,
        feature_flags: Optional[Dict[str, bool]] = None,
        compiler_options: Optional[Dict[str, Any]] = None,
        browser: Optional[str] = None,
        browser_options: Optional[Dict[str, Any]] = None,
        **kwargs
    ):
        """
        Initialize a hardware profile with specific configuration.
        
        Args:
            backend: Hardware backend name (e.g., "cuda", "openvino", "webgpu")
            device_id: Specific device ID or name (e.g., 0 for cuda:0)
            memory_limit: Maximum memory usage (e.g., "4GB")
            precision: Computation precision (e.g., "fp32", "fp16", "int8", "auto")
            optimization_level: Overall optimization strategy
            quantization: Quantization-specific configuration
            feature_flags: Enable/disable specific hardware features
            compiler_options: Backend-specific compiler options
            browser: Browser name for WebNN/WebGPU backends (e.g., "chrome", "firefox")
            browser_options: Browser-specific configuration options
            **kwargs: Additional backend-specific options
        """
        self.backend = backend
        self.device_id = device_id
        self.memory_limit = memory_limit
        self.precision = precision
        self.optimization_level = optimization_level
        self.quantization = quantization or {}
        self.feature_flags = feature_flags or {}
        self.compiler_options = compiler_options or {}
        self.browser = browser
        self.browser_options = browser_options or {}
        self.extra_options = kwargs
        
        # Map legacy backend names to standardized names
        self._normalize_backend_name()
        
    def _normalize_backend_name(self):
        """Normalize backend name to standard format."""
        backend_mapping = {
            "gpu": "cuda",
            "nvidia": "cuda",
            "amd": "rocm",
            "apple": "mps",
            "intel": "openvino",
            "qnn": "qualcomm",
            "snapdragon": "qualcomm",
            "web": "webgpu"
        }
        
        self.backend = backend_mapping.get(self.backend.lower(), self.backend.lower())
        
    def to_dict(self) -> Dict[str, Any]:
        """Convert hardware profile to dictionary format."""
        return {
            "backend": self.backend,
            "device_id": self.device_id,
            "memory_limit": self.memory_limit,
            "precision": self.precision,
            "optimization_level": self.optimization_level,
            "quantization": self.quantization,
            "feature_flags": self.feature_flags,
            "compiler_options": self.compiler_options,
            "browser": self.browser,
            "browser_options": self.browser_options,
            **self.extra_options
        }
        
    @classmethod
    def from_dict(cls, config_dict: Dict[str, Any]) -> 'HardwareProfile':
        """Create hardware profile from dictionary configuration."""
        return cls(**config_dict)
    
    def __repr__(self) -> str:
        """String representation of hardware profile."""
        return f"HardwareProfile(backend={self.backend}, device_id={self.device_id}, precision={self.precision})"
    
    def get_worker_compatible_config(self) -> Dict[str, Any]:
        """
        Get configuration compatible with the worker architecture.
        
        This method converts the hardware profile to a format compatible
        with the existing worker implementation for backward compatibility.
        """
        worker_config = {
            "hardware_type": self.backend,
            "device_id": self.device_id,
        }
        
        # Map precision to worker format
        if self.precision != "auto":
            worker_config["precision"] = self.precision
            
        # Add memory limit if specified
        if self.memory_limit:
            worker_config["memory_limit"] = self.memory_limit
            
        # Add browser configuration for web backends
        if self.backend in ["webgpu", "webnn"] and self.browser:
            worker_config["browser"] = self.browser
            worker_config["browser_options"] = self.browser_options
            
        # Add optimization level
        if self.optimization_level != "default":
            worker_config["optimization_level"] = self.optimization_level
            
        # Add feature flags
        if self.feature_flags:
            for flag, value in self.feature_flags.items():
                worker_config[flag] = value
                
        return worker_config

File: ipfs_accelerate_js/src/utils/hardware_profile.ts
// FI: any;
 * Convert: any;
 * Conversi: any;
 * Th: any;
 * Conversi: any;
 */;

import { HardwareAbstract: any;

// WebG: any;
export interface Props {memory_limit: worker_co: any;
  feature_fl: any;}

/** Hardwa: any;

Th: any;
th: any;

class $1 extends $2 {/** Configurati: any;
  backen: any;
  
  function __init__(): any:  any: any) { any {: any {) { a: an: any; t: any;
  $1) { string: any: any: any: any: any: any = "auto",;"
  $1: $2 = 0: a: any;
  memory_limit: Union[int, str | null] = nu: any;
  $1: string: any: any: any: any: any: any = "auto",;"
  optimization_level: Literal["default", "performance", "memory", "balanced"] = "default",;"
  quantization: Record<str, Any | null> = nu: any;
  feature_flags: Record<str, bool | null> = nu: any;
  compiler_options: Record<str, Any | null> = nu: any;
  $1: $2 | null: any: any: any = nu: any;
  browser_options: Record<str, Any | null> = nu: any;
  **kwargs;
  ):;
    /** Initiali: any;
    
    A: any;
      back: any;
      device_id: Specific device ID || name ()e.g., 0 for ((((((cuda) { any) {0);
      memory_limit) { Maximum) { an) { an: any;
      precisi) { an: any;
      optimization_le: any;
      quantizat: any;
      feature_fl: any;
      compiler_opti: any;
      browser: Browser name for ((((((WebNN/WebGPU backends () {e.g., "chrome", "firefox");"
      browser_options) { Browser) { an) { an: any;
      **kwargs) { Additiona) { an: any;
      this.backend = back: any;
      this.device_id = device: any;
      this.memory_limit = memory_li: any;
      this.precision = precis: any;
      this.optimization_level = optimization_le: any;
      this.quantization = quantization || {}
      this.feature_flags = feature_flags || {}
      this.compiler_options = compiler_options || {}
      this.browser = brow: any;
      this.browser_options = browser_options || {}
      this.extra_options = kwa: any;
    
    // M: any;
      th: any;
    ;
  $1($2) {
    /** Normali: any;
    backend_mapping) { any: any = {}
    "gpu": "cuda",;"
    "nvidia": "cuda",;"
    "amd": "rocm",;"
    "apple": "mps",;"
    "intel": "openvino",;"
    "qnn": "qualcomm",;"
    "snapdragon": "qualcomm",;"
    "web": "webgpu";"
    }
    this.backend = backend_mappi: any;
    
    functi: any;
    /** Conve: any;
      return {}
      "backend": th: any;"
      "device_id": th: any;"
      "memory_limit": th: any;"
      "precision": th: any;"
      "optimization_level": th: any;"
      "quantization": th: any;"
      "feature_flags": th: any;"
      "compiler_options": th: any;"
      "browser": th: any;"
      "browser_options": th: any;"
      **this.extra_options;
      }
    
      @classmethod;
      functi: any;
      /** Crea: any;
    retu: any;
  
  $1($2): $3 {/** Stri: any;
    retu: any;
    /** G: any;
    
    Th: any;
    wi: any;
    worker_config) { any) { any: any: any: any: any = {}
    "hardware_type") {this.backend,;"
    "device_id": th: any;"
    if ((((($1) {
      worker_config["precision"] = this) { an) { an: any;"
      ,;
    // Add memory limit if ((($1) {
    if ($1) {worker_config["memory_limit"] = this) { an) { an: any;"
      ,;
    // Add browser configuration for ((web backends}
      if ((($1) {,;
      worker_config["browser"] = this) { an) { an: any;"
      worker_config["browser_options"] = this) { an) { an: any;"
      ,;
    // Add optimization level}
    if ((($1) {worker_config["optimization_level"] = this) { an) { an: any;"
      ,;
    // Add feature flags}
    if (($1) {
      for (flag, value in this.Object.entries($1))) {
        worker_config[flag] = value) { an) { an) { an: any;